plugins/channel-safely.cpp

```diff
@@ -0,0 +1,480 @@
+        df::coord local(position);
+        local.x = local.x % 16;
+        local.y = local.y % 16;
+        df::coord above(position);
+        above.z++;
+        df::map_block* block = Maps::getTileBlock(position);
+        //activate designation if group is done now, postpone if not
+        manager.manage_safety(out, block, local, position, above);
+    }
+}
+
+void cancelJob(df::job* job) {
+    if(job != nullptr) {
+        df::coord &pos = job->pos;
+        df::map_block *job_block = Maps::getTileBlock(pos);
+        uint16_t x, y;
+        x = pos.x % 16;
+        y = pos.y % 16;
+        df::tile_designation &designation = job_block->designation[x][y];
+        designation.bits.dig = is_channel(job) ?
+                               df::tile_dig_designation::Channel : df::tile_dig_designation::Default;
+        Job::removeJob(job);
+    }
+}
+
+bool is_group_done(const GroupData::Group &group){
+    return group.empty();
+}
+
+bool is_dig(df::job* job) {
+    return job->job_type == df::job_type::Dig;
+}
+
+bool is_channel(df::job* job){
+    return job->job_type == df::job_type::DigChannel;
+}
+
+bool is_channel(df::tile_designation &designation){
+    return designation.bits.dig == df::tile_dig_designation::Channel;
+}
+
+void GroupData::read(bool full_scan){
+    if(full_scan){
+        free_spots.clear();
+        groups.clear();
+        groups_map.clear();
+    }
+    foreach_block(full_scan);
+    jobs.read();
+    for(auto &map_entry : jobs){
+        df::job* job = map_entry.second;
+        if(is_channel(job)){
+            add(map_entry.first, Maps::getTileBlock(job->pos));
+        }
+    }
+}
+
+void GroupData::foreach_block(bool full_scan) {
+    static std::once_flag getMapSize;
+    static uint32_t x, y, z;
+    std::call_once(getMapSize, Maps::getSize, x, y, z);
+    if(debug_out) debug_out->print("map size: %d, %d, %d\n", x,y,z);
+    if(debug_out) debug_out->print("full_scan: %s\n", full_scan ? "true" : "false");
+    for (int ix = 0; ix < x; ++ix) {
+        for (int iy = 0; iy < y; ++iy) {
+            for (int iz = z - 1; iz >= 0; --iz) {
+                df::map_block* block = Maps::getBlock(ix, iy, iz);
+                if(full_scan || block->flags.bits.designated) {
+                    foreach_tile(block, iz);
+                }
+            }
+        }
+    }
+}
+
+void GroupData::foreach_tile(df::map_block* block, int z) {
+    for (int16_t local_x = 0; local_x < 16; ++local_x) {
+        for (int16_t local_y = 0; local_y < 16; ++local_y) {
+            if(is_channel(block->designation[local_x][local_y])){
+                df::coord world_pos(block->map_pos);
+                world_pos.x += local_x;
+                world_pos.y += local_y;
+                world_pos.z = z;
+                add(world_pos, block);
+            }
+        }
+    }
+}
+
+void GroupData::add(df::coord world_pos, df::map_block* block) {
+    if(groups_map.find(world_pos) == groups_map.end()) {
+        df::coord neighbors[8];
```

Comment:
Might be worth having a helper function to convert a `df::coord` into its neighbors, to avoid having to duplicate this segment from `manageNeighbours()` above